driver-core/src/main/com/mongodb/connection/DefaultAuthenticator.java

```diff
@@ -16,33 +16,95 @@
     }
 
     @Override
     void authenticate(final InternalConnection connection, final ConnectionDescription connectionDescription) {
-        createAuthenticator(connectionDescription).authenticate(connection, connectionDescription);
+        Authenticator authenticator;
+        if (connectionDescription.getServerVersion().compareTo(FOUR_ZERO) < 0) {
+            authenticator = getLegacyDefaultAuthenticator(connectionDescription);
+        } else {
+            try {
+                BsonDocument isMasterResult = executeCommand("admin", createIsMasterCommand(), connection);
+                authenticator = getAuthenticatorFromIsMasterResult(isMasterResult);
+            } catch (Exception e) {
+                throw wrapException(e);
+            }
+        }
+        authenticator.authenticate(connection, connectionDescription);
     }
 
     @Override
     void authenticateAsync(final InternalConnection connection, final ConnectionDescription connectionDescription,
                            final SingleResultCallback<Void> callback) {
-        createAuthenticator(connectionDescription).authenticateAsync(connection, connectionDescription, callback);
+        if (connectionDescription.getServerVersion().compareTo(FOUR_ZERO) < 0) {
+            getLegacyDefaultAuthenticator(connectionDescription).authenticateAsync(connection, connectionDescription, callback);
+        } else {
+            executeCommandAsync("admin", createIsMasterCommand(), connection, new SingleResultCallback<BsonDocument>() {
+                @Override
+                public void onResult(final BsonDocument result, final Throwable t) {
+                    if (t != null) {
+                        callback.onResult(null, wrapException(t));
+                    } else {
+                        getAuthenticatorFromIsMasterResult(result).authenticateAsync(connection, connectionDescription, callback);
+                    }
+                }
+            });
+        }
+    }
+
+    private Authenticator getAuthenticatorFromIsMasterResult(final BsonDocument isMasterResult) {
+        BsonArray saslSupportedMechs = isMasterResult.getArray("saslSupportedMechs", DEFAULT_SUPPORTED_MECHANISMS);
+        AuthenticationMechanism mechanism = saslSupportedMechs.contains(DEFAULT_MECHANISM_NAME) ? SCRAM_SHA_256 : SCRAM_SHA_1;
+        return new ScramShaAuthenticator(getCredential(), mechanism);
     }
 
-    Authenticator createAuthenticator(final ConnectionDescription connectionDescription) {
-        if (connectionDescription.getServerVersion().compareTo(new ServerVersion(2, 7)) >= 0) {
-            return new ScramSha1Authenticator(getCredential());
+    private Authenticator getLegacyDefaultAuthenticator(final ConnectionDescription connectionDescription) {
+        if (connectionDescription.getServerVersion().compareTo(THREE_ZERO) >= 0) {
+            return new ScramShaAuthenticator(getCredential(), SCRAM_SHA_1);
```

Comment:
I find it confusing to pass in a MongoCredential and an AuthenticationMechanism, since a MongoCredential has an AuthenticationMechanism property.  How about something like this instead:

    return new ScramShaAuthenticator(getCredential().withMechanism(SCRAM_SHA_1));

Reply:
Its an artifact of how it was in the past. Will improve.